src/Microsoft.OData.Core/ODataWriterCore.cs

```diff
@@ -2834,6 +2845,32 @@ private ODataPath AppendEntitySetKeySegment(ODataPath odataPath, bool throwIfFai
             return path;
         }
 
+        private bool TryBuildKeySegment(out KeySegment keySegment)

```

Comment:
>TryBuildKeySegment [](http://example.com/codeflow?start=21&length=18)

This seems to be the same logic as in AppendKeySegment for getting the key segment, except that AppendKeySegment returns the path with the key segment appended.  Can AppendKeySegment call this method to get the key segment in order to avoid duplicating the code in two places?

Reply:
Another reference (only use) for 'AppendEntitySetKeySegment' is for a nested complex property. I think it could be related to the nested navigation property. If it's ok, we 'd need to do the same thing for complex property? Any thougths?

By the way, i have to create another private method to include the duplicate codes.